// telaug/telemetry/telemetryStrings.js
"use strict";
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//
Object.defineProperty(exports, "__esModule", { value: true });
class TelemetryPropertyNames {
    static get FAULT_TYPE() {
        return TelemetryPropertyNames.FEATURE_NAME + 'FaultType';
    }
    static get CORRELATION_ID() {
        return TelemetryPropertyNames.FEATURE_NAME + 'CorrelationId';
    }
    static get LOG_LINES() {
        return TelemetryPropertyNames.FEATURE_NAME + 'LogLines';
    }
    static get EVENT_MESSAGE() {
        return TelemetryPropertyNames.FEATURE_NAME + 'Message';
    }
    static get EVENT_EXCEPTION_STACK() {
        return TelemetryPropertyNames.FEATURE_NAME + 'ExceptionStack';
    }
    static get EVENT_RESULT() {
        return TelemetryPropertyNames.FEATURE_NAME + 'Result';
    }
    static get EVENT_DURATION() {
        return TelemetryPropertyNames.FEATURE_NAME + 'Duration';
    }
    static get SESSION_ID() {
        return TelemetryPropertyNames.FEATURE_NAME + 'SessionId';
    }
}
TelemetryPropertyNames.FEATURE_NAME = '';
exports.TelemetryPropertyNames = TelemetryPropertyNames;
