// vscode-kubernetes-tools-api/js/helper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = __importStar(require("vscode"));
const constants_1 = require("./constants");
class Lazy {
    constructor(fn) {
        this.fn = fn;
        this.value = undefined;
    }
    get() {
        if (this.value === undefined) {
            this.value = this.fn();
        }
        return this.value;
    }
}
class ExtensionHelper {
    constructor() {
        this.apiBroker = new Lazy(() => getBroker());
        this.clusterProvider = readonlify({
            v1: this.get("clusterprovider", "v1"),
        });
        this.kubectl = readonlify({
            v1: this.get("kubectl", "v1"),
        });
        this.clusterExplorer = readonlify({
            v1: this.get("clusterexplorer", "v1"),
        });
        this.helm = readonlify({
            v1: this.get("helm", "v1"),
        });
        this.cloudExplorer = readonlify({
            v1: this.get("cloudexplorer", "v1"),
        });
    }
    getCore(component, version) {
        return __awaiter(this, void 0, void 0, function* () {
            const apiBroker = yield this.apiBroker.get();
            if (!apiBroker) {
                return { available: false, reason: "extension-not-available" };
            }
            if (!apiBroker.get) {
                return { available: false, reason: "version-unknown" };
            }
            return apiBroker.get(component, version);
        });
    }
    get(component, version) {
        return __awaiter(this, void 0, void 0, function* () {
            const api = yield this.getCore(component, version);
            return api;
        });
    }
}
exports.ExtensionHelper = ExtensionHelper;
function readonlify(t) {
    return t;
}
function getBroker() {
    return __awaiter(this, void 0, void 0, function* () {
        const extension = vscode.extensions.getExtension(constants_1.MS_KUBERNETES_EXTENSION_ID);
        if (!extension) {
            return undefined;
        }
        const apiBroker = yield extension.activate();
        return apiBroker;
    });
}
